src/transformers/modeling_utils.py

```diff
@@ -1371,11 +1373,53 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                             else:
                                 raise ValueError from e
                     except (UnicodeDecodeError, ValueError):
-                        raise OSError(
-                            f"Unable to load weights from pytorch checkpoint file for '{pretrained_model_name_or_path}' "
-                            f"at '{resolved_archive_file}'. "
-                            "If you tried to load a PyTorch model from a TF 2.0 checkpoint, please set from_tf=True."
-                        )
+                        # Check sha256
+                        def sha256(filename, block_size=4096 * 1024):
+                            # Python program to find SHA256 hash string of a file
+
+                            sha256_hash = hashlib.sha256()
+                            with open(filename, "rb") as f:
+                                # Read and update hash string value in blocks of 4Mo
+                                for byte_block in iter(lambda: f.read(block_size), b""):
+                                    sha256_hash.update(byte_block)
+
+                            return sha256_hash.hexdigest()
+
+                        actual_sha = sha256(resolved_archive_file)
+                        metafilename = f"{resolved_archive_file}.json"
+                        try:
+                            with open(metafilename, "r", encoding="utf-8") as f:
+                                data = json.load(f)
+                                # For some reason etag contains quotes, we just want the sha.
+                                expected_sha = data["etag"][1:-1]
+                        except Exception:
+                            expected_sha = None
+
+                        if actual_sha != expected_sha:
+                            logger.warning(
+                                f"Look like a corrupted file is on disk ({actual_sha}!={expected_sha}), attempting to recover by downloading again."
+                            )
+
+                            # Remove corrupted file
+                            os.remove(resolved_archive_file)
+                            # Redownload file
+                            resolved_archive_file = cached_path(
+                                archive_file,
+                                cache_dir=cache_dir,
+                                force_download=force_download,
+                                proxies=proxies,
+                                resume_download=resume_download,
+                                local_files_only=local_files_only,
+                                use_auth_token=use_auth_token,
+                                user_agent=user_agent,
+                            )
+                            state_dict = torch.load(resolved_archive_file, map_location="cpu")
```

Comment:
Instead of redownloading here, could we remove the cached file and instruct the user in re-launching the instruction? You would need to re-put the exceptions chained here in a case of connection errors.

Reply:
Makes sense, I'll try to keep the redownload logic but if it proves too hard or too much of a big modification, I'll just raise an error instructing user to restart.